### PR TITLE
Move findBinaryFinder and isFunctionEnabled away from OC_Helper

### PR DESCRIPTION
**Pull request by CarlSchwan**
findBinaryFinder is now a service that is still private but with some
minor optimization (remove the hasKey check).

isFunctionEnabled is now in OCP\Util

Both functions are still kept but all internal usage in nextcloud/server
were migrated to the new usage, so we can remove it in 26.

Before: 2 redis cache requests either hasKey and get or hasKey and set

image

Now: 1 redis cache request (get) or 2 redis cache requests (get and set)

image

### CODE REVIEW

**Review by blizzz — APPROVED** [review]

File: lib/private/legacy/OC_Helper.php [hunk 1]
```diff
@@ -436,45 +436,19 @@ public static function uploadLimit() {
 	 *
 	 * @param string $function_name
 	 * @return bool
+	 * @deprecated Since 25.0.0 use \OCP\Util::isFunctionEnabled instead
 	 */
 	public static function is_function_enabled($function_name) {
```
Comment (nickvergessen) [inline lib/private/legacy/OC_Helper.php:441]:
```suggestion
	public static function is_function_enabled(string $function_name): bool {
```

File: lib/private/BinaryFinder.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+namespace OC;
+
+use OCP\ICache;
+use OCP\ICacheFactory;
+use Symfony\Component\Process\ExecutableFinder;
+
+/**
+ * Service that find the binary path for a program
+ */
+class BinaryFinder {
```
Comment (nickvergessen) [inline lib/private/BinaryFinder.php:32]: OCP interface ?

File: tests/lib/LargeFileHelperGetFileSizeTest.php [hunk 3]
```diff
@@ -63,7 +63,7 @@ public function testGetFileSizeViaExec($filename, $fileSize) {
 		if (escapeshellarg('strängé') !== '\'strängé\'') {
 			$this->markTestSkipped('Your escapeshell args removes accents');
 		}
-		if (!\OC_Helper::is_function_enabled('exec')) {
+		if (!\OC\Util::isFunctionEnabled('exec')) {
```
Comment (come-nc) [inline tests/lib/LargeFileHelperGetFileSizeTest.php:66]:
```suggestion
		if (!\OCP\Util::isFunctionEnabled('exec')) {
```

**Review by come-nc — CHANGES_REQUESTED** [review]

File: lib/private/legacy/OC_Helper.php [hunk 4]
```diff
@@ -434,47 +434,19 @@
...
-		if (in_array($function_name, $disabled)) {
-			return false;
-		}
-		$disabled = explode(',', $ini->get('suhosin.executor.func.blacklist') ?: '');
-		$disabled = array_map('trim', $disabled);
-		if (in_array($function_name, $disabled)) {
-			return false;
-		}
-		return true;
+	public static function is_function_enabled(string $function_name): booo {
```
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply lib/private/legacy/OC_Helper.php]: :ghost:

**Review by ChristophWurst — APPROVED** [review]

File: lib/private/Mail/Mailer.php [hunk 5]
```diff
@@ -309,7 +305,7 @@ protected function getSendMailInstance(): \Swift_SendmailTransport {
 				$binaryPath = '/var/qmail/bin/sendmail';
 				break;
 			default:
-				$sendmail = \OC_Helper::findBinaryPath('sendmail');
+				$sendmail = \OCP\Server::get(BinaryFinder::class)->findBinaryPath('sendmail');
```
Comment (come-nc) [inline lib/private/Mail/Mailer.php:308]:
```suggestion
				$sendmail = \OCP\Server::get(IBinaryFinder::class)->findBinaryPath('sendmail');
```

File: lib/private/Server.php [hunk 6]
```diff
@@ -335,7 +336,8 @@ public function __construct($webRoot, \OC\Config $config) {
 				$c->get(GeneratorHelper::class),
 				$c->get(ISession::class)->get('user_id'),
 				$c->get(Coordinator::class),
-				$c->get(IServerContainer::class)
+				$c->get(IServerContainer::class),
+				$c->get(BinaryFinder::class)
```
Comment (come-nc) [inline lib/private/Server.php:340]:
```suggestion
				$c->get(IBinaryFinder::class)
```

**Review by come-nc — APPROVED** [review]

File: lib/private/Mail/Mailer.php [hunk 7]
```diff
@@ -309,7 +305,7 @@ protected function getSendMailInstance(): \Swift_SendmailTransport {
 				$binaryPath = '/var/qmail/bin/sendmail';
 				break;
 			default:
-				$sendmail = \OC_Helper::findBinaryPath('sendmail');
+				$sendmail = \OCP\Server::get(IBinaryFinder::class)->findBinaryPath('sendmail');
```
Comment (nickvergessen) [inline lib/private/Mail/Mailer.php:308]: DI?

**Review by nickvergessen — APPROVED** [review]

File: lib/private/Mail/Mailer.php (hunk 7 above)
Comment (CarlSchwan) [reply lib/private/Mail/Mailer.php:308]: I tried but that involves quite a lot of change a bit everywhere

**Pull request merged**